edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/upstream/BrokeredCloudConnectionProvider.cs

```diff
@@ -27,10 +27,33 @@ public Task<Try<ICloudConnection>> Connect(IClientCredentials clientCredentials,
             return this.Connect(clientCredentials.Identity, connectionStatusChangedHandler);
         }
 
-        public Task<Try<ICloudConnection>> Connect(IIdentity identity, Action<string, CloudConnectionStatus> connectionStatusChangedHandler)
+        public async Task<Try<ICloudConnection>> Connect(IIdentity identity, Action<string, CloudConnectionStatus> connectionStatusChangedHandler)
         {
+            if (!await this.IsConnected())
+            {
+                return new Try<ICloudConnection>(new Exception("Bridge is not connected upstream"));
+            }
+
             var cloudProxy = new BrokeredCloudProxy(identity, this.cloudProxyDispatcher, connectionStatusChangedHandler);
-            return Task.FromResult(new Try<ICloudConnection>(new BrokeredCloudConnection(cloudProxy)));
+            return new Try<ICloudConnection>(new BrokeredCloudConnection(cloudProxy));
+        }
+
+        async Task<bool> IsConnected()
```

Comment:
can an event be used to the dispacher that gets triggered when it connect? or if it only has to wait for a short time and then throw you can use the RetryStrategy from utils

Reply:
Changed to retry strategy. Using events seems more complex for me - normally it should be connected, so normally the tests pass (and the method returns true). As described above, this trying loop just mitigates the noise mainly at startup, when it can happen that edgehub-core has not received the connection event from the broker (from upstream bridge)